File: go/orm/internal/metadata/entity_map_builder_test.go
package metadata_test

import (
	"testing"

	"github.com/kosmassp/SimpleOrm/go/orm"
	"github.com/kosmassp/SimpleOrm/go/orm/internal/core"
	"github.com/kosmassp/SimpleOrm/go/orm/internal/metadata"
	"github.com/kosmassp/SimpleOrm/go/orm/sample"
)

// The manual loader and the loader precedence (explicit → declaration →
// convention). Mirrors dotnet/tests/SimpleOrm.Tests/EntityMapBuilderTests.cs.

// legacyFixture deliberately carries no orm tags: the "can't or won't annotate" case.
type legacyFixture struct {
	ID          int64
	DisplayName *string
}

// personFixture is a plain unannotated type for the convention loader.
type personFixture struct {
	ID        int64
	FirstName *string
}

func TestBuilder_MapsOnlyDeclaredPropertiesWithExplicitNames(t *testing.T) {
	builder := metadata.NewEntityMapBuilder[legacyFixture]().ToTable("legacy_items")
	builder.Property("ID").Column("legacy_id").Key().Generated()
	builder.Property("DisplayName").Column("display_name")

	options := (&orm.MappingOptions{}).Register(builder)
	m, err := metadata.Load[legacyFixture](metadata.NewLoader(options))
	if err != nil {
		t.Fatal(err)
	}

	if m.RelationName != "legacy_items" {
		t.Errorf("RelationName = %q", m.RelationName)
	}
	if m.KeyStrategy != core.KeyDatabaseGenerated {
		t.Errorf("KeyStrategy = %v", m.KeyStrategy)
	}
	if !equalStrings(columnNames(m), []string{"legacy_id", "display_name"}) {
		t.Errorf("columns = %v", columnNames(m))
	}
}

func TestExplicitRegistration_WinsOverDeclarations(t *testing.T) {
	builder := metadata.NewEntityMapBuilder[sample.User]().ToTable("users_manual")
	builder.Property("ID").Key().Generated()
	builder.Property("Name")

	options := (&orm.MappingOptions{}).Register(builder)
	m, err := metadata.Load[sample.User](metadata.NewLoader(options))
	if err != nil {
		t.Fatal(err)
	}

	if m.RelationName != "users_manual" {
		t.Errorf("RelationName = %q", m.RelationName)
	}
	if len(m.Properties) != 2 {
		t.Errorf("Properties = %+v, want 2", m.Properties)
	}
}

func TestConventionLoader_MapsUnannotatedTypes(t *testing.T) {
	m, err := metadata.Load[personFixture](metadata.NewLoader(nil))
	if err != nil {
		t.Fatal(err)
	}

	if m.Kind != core.RelationTable {
		t.Errorf("Kind = %v, want Table", m.Kind)
	}
	if m.RelationName != "person_fixture" {
		t.Errorf("RelationName = %q, want person_fixture", m.RelationName)
	}
	if m.KeyStrategy != core.KeyDatabaseGenerated {
		t.Errorf("KeyStrategy = %v, want DatabaseGenerated", m.KeyStrategy)
	}
	if !equalStrings(columnNames(m), []string{"id", "first_name"}) {
		t.Errorf("columns = %v", columnNames(m))
	}
	if !propertyByColumn(m, "first_name").IsNullable {
		t.Error("first_name must be nullable (pointer field)")
	}
}

func TestBuilder_FailuresCarryCodesToo(t *testing.T) {
	builder := metadata.NewEntityMapBuilder[legacyFixture]()
	builder.Property("ID").Column("same")
	builder.Property("DisplayName").Column("same").Key()

	options := (&orm.MappingOptions{}).Register(builder)
	_, err := metadata.Load[legacyFixture](metadata.NewLoader(options))
	if !core.HasCode(err, "MAP-018") {
		t.Fatalf("expected MAP-018, got %v", err)
	}
}

func TestBuilder_UnknownFieldNameIsMAP023(t *testing.T) {
	builder := metadata.NewEntityMapBuilder[legacyFixture]()
	builder.Property("NoSuchField").Column("x")

	options := (&orm.MappingOptions{}).Register(builder)
	_, err := metadata.Load[legacyFixture](metadata.NewLoader(options))
	if !core.HasCode(err, "MAP-023") {
		t.Fatalf("expected MAP-023, got %v", err)
	}
}

File: php/tests/Metadata/EntityMapBuilderTest.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Tests\Metadata;

use PHPUnit\Framework\Attributes\Test;
use PHPUnit\Framework\TestCase;
use SimpleOrm\Errors\MappingException;
use SimpleOrm\Metadata\EntityMapBuilder;
use SimpleOrm\Metadata\KeyStrategy;

/** The manual fluent builder (§7.2): maps only declared properties, runs the same validations as the attribute loader. */
final class EntityMapBuilderTest extends TestCase
{
    #[Test]
    public function builder_maps_only_declared_properties_with_explicit_names(): void
    {
        $map = EntityMapBuilder::for(Legacy::class)
            ->table('legacy_items')
            ->column('id', name: 'legacy_id', key: true, generated: true)
            ->column('displayName', name: 'display_name')
            ->build();

        self::assertSame('legacy_items', $map->relationName);
        self::assertSame(KeyStrategy::DatabaseGenerated, $map->keyStrategy);
        self::assertSame(['legacy_id', 'display_name'], array_map(static fn ($p) => $p->columnName, $map->properties));
    }

    #[Test]
    public function table_name_falls_back_to_the_naming_convention_when_never_set(): void
    {
        $map = EntityMapBuilder::for(Legacy::class)
            ->column('id', key: true, generated: true)
            ->build();

        self::assertSame('legacy', $map->relationName);
    }

    #[Test]
    public function builder_failures_carry_codes_too(): void
    {
        $builder = EntityMapBuilder::for(Legacy::class)
            ->column('id', name: 'same')
            ->column('displayName', name: 'same', key: true);

        try {
            $builder->build();
            self::fail('expected a MappingException');
        } catch (MappingException $e) {
            self::assertContains('MAP-018', array_map(static fn ($error) => $error->code, $e->errors));
        }
    }
}

// Deliberately no SimpleOrm attributes: the "can't or won't annotate" case.
final class Legacy
{
    public int $id;

    public ?string $displayName = null;
}
